src/graph/executor/query/ExpandAllExecutor.cpp

```diff
@@ -0,0 +1,379 @@
+        curLimit_ = 0;
+        curMaxLimit_ = stepLimits_.empty() ? std::numeric_limits<int64_t>::max()
+                                           : stepLimits_[currentStep_ - 2];
+        return handleResponse(std::move(resp)).ensure([this, expandTime]() {
+          std::string timeName = "graphExpandExpandTime+" + folly::to<std::string>(currentStep_);
+          otherStats_.emplace(timeName, folly::sformat("{}(us)", expandTime.elapsedInUSec()));
+        });
+      })
+      .thenValue([this](Status s) -> folly::Future<Status> {
+        NG_RETURN_IF_ERROR(s);
+        if (qctx()->isKilled()) {
+          return Status::OK();
+        }
+        if (currentStep_ <= maxSteps_) {
+          if (!nextStepVids_.empty()) {
+            return getNeighbors();
+          } else if (!preVisitedVids_.empty()) {
+            return expandFromCache();
+          } else {
+            return finish(ResultBuilder().value(Value(std::move(result_))).build());
+          }
+        } else {
+          return finish(ResultBuilder().value(Value(std::move(result_))).build());
+        }
+      });
+}
+
+folly::Future<Status> ExpandAllExecutor::expandFromCache() {
+  for (; currentStep_ <= maxSteps_; ++currentStep_) {
+    time::Duration expandTime;
+    if (qctx()->isKilled()) {
+      return Status::OK();
+    }
+    curLimit_ = 0;
+    curMaxLimit_ =
+        stepLimits_.empty() ? std::numeric_limits<int64_t>::max() : stepLimits_[currentStep_ - 2];
+
+    std::vector<int64_t> samples;
+    if (sample_) {
+      int64_t size = 0;
+      for (auto& vid : preVisitedVids_) {
+        size += adjList_[vid].size();
+      }
+      algorithm::ReservoirSampling<int64_t> sampler(curMaxLimit_, size);
+      samples = sampler.samples();
+      std::sort(samples.begin(), samples.end(), [](int64_t a, int64_t b) { return a > b; });
+    }
+
+    std::unordered_map<Value, std::unordered_set<Value>> dst2VidsMap;
+    std::unordered_set<Value> visitedVids;
+    getNeighborsFromCache(dst2VidsMap, visitedVids, samples);
+    preVisitedVids_.swap(visitedVids);
+    preDst2VidsMap_.swap(dst2VidsMap);
+    std::string timeName = "graphCacheExpandTime+" + folly::to<std::string>(currentStep_);
+    otherStats_.emplace(timeName, folly::sformat("{}(us)", expandTime.elapsedInUSec()));
+    if (!nextStepVids_.empty()) {
+      return getNeighbors();
+    }
+  }
+  return finish(ResultBuilder().value(Value(std::move(result_))).build());
+}
+
+void ExpandAllExecutor::getNeighborsFromCache(
+    std::unordered_map<Value, std::unordered_set<Value>>& dst2VidsMap,
+    std::unordered_set<Value>& visitedVids,
+    std::vector<int64_t>& samples) {
+  for (const auto& vid : preVisitedVids_) {
+    auto findVid = adjList_.find(vid);
+    if (findVid == adjList_.end()) {
+      continue;
+    }
+    auto& adjEdgeProps = findVid->second;
+    auto& vertexProps = adjEdgeProps.back();
+    for (auto edgeIter = adjEdgeProps.begin(); edgeIter != adjEdgeProps.end() - 1; ++edgeIter) {
```

Comment:
edgeIter < adjEdgeProps.end()

Reply:
should be edgeIter < adjEdgeProps.end() - 1 ,  and  edgeIter != adjEdgeProps.end() -1 is ok